alf/algorithms/distributed_off_policy_algorithm.py

```diff
@@ -0,0 +1,541 @@
+        # start the data receiver subprocess
+        process = mp.Process(
+            target=receive_experience_data,
+            args=(self._replay_buffer, max(0,
+                                           PerProcessContext().ddp_rank),
+                  self._unroller_ip, self._unroller_port,
+                  self._total_exps_received),
+            daemon=True)
+        process.start()
+
+    def _send_trainer_info_to_unroller(self):
+        """Create a REP socket and send the number of workers to the unroller,
+        so that the unroller is able to know the worker ids
+
+        'worker-0', 'worker-1', ...
+
+        to route experience data to.
+
+        Also send the trainer's ip and port to the unroller.
+        """
+        socket, cxt = create_zmq_socket(zmq.REQ, self._unroller_ip,
+                                        self._unroller_port)
+        logging.info(
+            '[worker-0] Waiting to send the unroller the number of workers...')
+        trainer_ip = get_local_ip()
+        socket.send_string(
+            f'worker-0: {PerProcessContext().num_processes}, {trainer_ip}, {self._port}'
+        )
+        message = socket.recv_string()
+        assert message == 'unroller: ok'
+        socket.close()
+        cxt.term()
+
+    @property
+    def utd(self):
+        with self._total_exps_received.get_lock():
+            exps_received = self._total_exps_received.value
+        if exps_received == 0:
+            return 0
+        return self._total_updates / exps_received
+
+    def _train_iter_off_policy(self):
+        # A worker will pause when either happens:
+        # 1. replay buffer is not ready (initial collect steps not reached)
+        # 2. utd ratio is too high (training is too fast; wait for more data)
+        while True:
+            replay_buffer_not_ready = (self._replay_buffer.total_size <
+                                       self._config.initial_collect_steps)
+            utd_exceeded = self.utd > self._max_utd_ratio
+            if not (replay_buffer_not_ready or utd_exceeded):
+                break
+            if replay_buffer_not_ready:
+                logging.debug(
+                    f"[worker-{PerProcessContext().ddp_rank}] Pause: replay buffer is not ready yet"
+                )
+            if utd_exceeded:
+                logging.debug(
+                    f"[worker-{PerProcessContext().ddp_rank}] Pause: UTD exceeded"
+                )
+            time.sleep(0.1)
+
+        steps = super()._train_iter_off_policy()
+        self._total_updates += self._config.num_updates_per_train_iter
+        return steps
+
+    def _backward_and_gradient_update(self, loss):
+        if self.is_main_ddp_rank:
+            # Lock the params to avoid sending them to the unroller while being updated.
+            # We only need to lock for the main DDP rank because the other DDP ranks
+            # will sync with it.
+            with self._params_lock:
```

Comment:
> Instead of allowing sending parameters at anytime, it is more reliable to only send at the end of an iteration.

I tried this method in the first place. But there is something tricky here. Imagine that rank 0 is responsible for sending the params to the unroller at an end of some training iter. When some other rank gets stuck due to `replay_buffer_not_ready` or `utd_exceeded`, it won't be able to perform backward. So rank 0 might also get stuck in DDP backward sync. If at this moment the unroller requests the param update, then the entire system will get a dead lock. 

I haven't figured out a way how to solve this issue except using a separate thread to respond to the unroller.

Reply:
> This is not reliable. There are other things besides parameters which can be changed during model training (e.g. various buffers). Instead of allowing sending parameters at anytime, it is more reliable to only send at the end of an iteration.

Now changed to trainer pushing params to unrollers and each unroller has a thread to receive params, without blocking unroll().